### DISCUSSION TITLE
Scrollable Jump Table (dynamic) question

### DISCUSSION THREAD

**Discussion by ViacheslavK9** [olikraus/u8g2#2333] [Q&A] [answered]
hi Oliver!

A great job you have done with your MUI extension! 

I have similar question to this (probably) https://github.com/olikraus/u8g2/discussions/2312

I am trying to create simple tree menu like

menu 10

--go to menu 11
--go to menu 12

menu 11

--go to menu 21
--go to menu 22
--go back

menu 12

--go to menu 23
--go to menu 24
--go back

I am using several Scrollable Jump Table (dynamic) like this:

 MUIF_U8G2_U16_LIST("ID", &selection, NULL, menu_get_str, menu_get_cnt, mui_u8g2_u16_list_goto_w1_pi),

 MUIF_U8G2_U16_LIST("AC", &selection2, NULL, menuAC_get_str, menu_get_cnt2, mui_u8g2_u16_list_goto_w1_pi).

...

My question is: when i go to the second menu e.g. 11 and choose there "go back" (to the main menu)
and then again select menu 11 my cursor is at the "go back" button (it remembers my previous choice (var selection2))
and i want it to "forget it" with condition "we came back to the main menu". 

I am trying to zero my selection2 variable, but it does not work.

//below code i do execute on going back to the main menu
selection2 = 0;
Serial.println(String(selection2));

also do check on a loop section, it shows zero but its not when i open menu 11.

Maybe you can say what is wrong.

Viacheslav

**Comment by ViacheslavK9** [comment]
sorry, maybe i ask not very clear, i try again:

if i am using Scrollable Jump Table (dynamic) like
 MUIF_U8G2_U16_LIST("ID", &selection, NULL, menu_get_str, menu_get_cnt, mui_u8g2_u16_list_goto_w1_pi)

how (and where) i can change variable "selection" by programming?

thanks!

**Reply by olikraus** [reply]
Hmm can you send a complete runnable example?

**Reply by ViacheslavK9** [reply]
Hi, yes, sure

``` C++
#include <Arduino.h>
#include <U8g2lib.h>
#include <MUIU8g2.h>
//#include <SimpleRotary.h>
#include <Versatile_RotaryEncoder.h>

// Set here your encoder reading pins
#define clk A2
#define dt A3
#define sw A4

// Create encoder object
Versatile_RotaryEncoder versatile_encoder(clk, dt, sw);


U8G2_SSD1306_128X64_NONAME_1_HW_I2C u8g2(U8G2_R2, /* reset=*/U8X8_PIN_NONE, /* clock=*/16, /* data=*/17);  // ESP32 Thing, HW I2C with pin remapping

MUIU8G2 mui;

uint8_t num_value = 0;
uint8_t bar_value = 0;
long LastTime;

uint16_t menu_get_cnt(void *data) {
  return 7;    /* number of menu entries */
}

uint16_t menu_get_cnt2(void *data) {
  return 5;    /* number of menu entries */
}

const char *menu_get_str(void *data, uint16_t index) {
  static const char *menu[] = { 
    MUI_11 "Device 1",
    MUI_11 "Device 2",
    MUI_11 "Device 3",
    MUI_11 "Device 4",
    MUI_11 "Device 5",
    MUI_11 "Device 6"
  };
  return menu[index];
}

const char *menuAC_get_str(void *data, uint16_t index) {
  static const char *menu[] = { 
    MUI_21 "Test",
    MUI_22 "Battery",
    MUI_23 "Timer",
    MUI_24 "Activation",
    MUI_25 "Back..",
  };
  return menu[index];
}

uint16_t selection = 0, selection2 = 0;

uint8_t mui_hrule(mui_t *ui, uint8_t msg) {
  if (msg == MUIF_MSG_DRAW) {
    u8g2.drawHLine(0, mui_get_y(ui), u8g2.getDisplayWidth());
    //Serial.println("drawHLine " + String(selection + 1));

    Serial.println(String(selection2));

    selection2 = 0;
  }
  return 0;
}

uint8_t mui_draw_current_device(mui_t *ui, uint8_t msg) {
  if (msg == MUIF_MSG_DRAW) {
    u8g2.setCursor(mui_get_x(ui ), mui_get_y(ui));

    u8g2.print("Device: " + String(selection + 1));

  }
  return 0;
}

//forced exit
uint8_t mui_go_main_menu(mui_t *ui, uint8_t msg) {
  if (msg == MUIF_MSG_FORM_START) {
    mui.sendSelectWithExecuteOnSelectFieldSearch();

    Serial.println("mui_go_main_menu");
  }
  return 0;
}

muif_t muif_list[] = {
  MUIF_U8G2_FONT_STYLE(0, u8g2_font_6x12_t_cyrillic), /* regular font */
  MUIF_U8G2_FONT_STYLE(1, u8g2_font_6x12_t_cyrillic), /* bold font */

  MUIF_VARIABLE("TM",&TestModeRadioChoice,mui_u8g2_u8_chkbox_wm_pi),

  MUIF_RO("HR", mui_hrule),
  MUIF_U8G2_LABEL(),
  MUIF_RO("GP", mui_u8g2_goto_data),
  MUIF_BUTTON("GC", mui_u8g2_goto_form_w1_pi),

  MUIF_U8G2_U16_LIST("ID", &selection, NULL, menu_get_str, menu_get_cnt, mui_u8g2_u16_list_goto_w1_pi),

  MUIF_U8G2_U16_LIST("AC", &selection2, NULL, menuAC_get_str, menu_get_cnt2, mui_u8g2_u16_list_goto_w1_pi),

  MUIF_RO("DD", mui_draw_current_device),

  MUIF_RO("GG", mui_go_main_menu),

  /* a button for the menu... */
  MUIF_EXECUTE_ON_SELECT_BUTTON("ST", mui_u8g2_btn_exit_wm_fi)  
  //MUIF_BUTTON("ST", mui_u8g2_btn_exit_wm_fi)   
};

fds_t fds_data[] =

  MUI_FORM(1)
  MUI_STYLE(1)
  MUI_LABEL(5, 10, "Choose:")
  MUI_XY("HR", 0,11)

  MUI_XYA("ID", 5, 24, 0)
  MUI_XYA("ID", 5, 36, 1)
  MUI_XYA("ID", 5, 48, 2)
  MUI_XYA("ID", 5, 60, 3)


  MUI_FORM(11)
  MUI_STYLE(1)
  MUI_XY("HR", 0,11)
  MUI_XY("DD", 5, 10)

  MUI_XYA("AC", 5, 24, 0)
  MUI_XYA("AC", 5, 36, 1)                                      
  MUI_XYA("AC", 5, 48, 2)
  MUI_XYA("AC", 5, 60, 3)

  MUI_FORM(25)
  MUI_STYLE(1)
  MUI_XYT("ST", 5, 24, "Exit")
  MUI_AUX("GG") 
;


uint8_t is_redraw = 1;
uint8_t rotate_event = 0;      // 0 = not turning, 1 = CW, 2 = CCW
uint8_t press_event = 0;       // 0 = not pushed, 1 = pushed
uint8_t long_press_event = 0;  // 0 = not pushed, 1 = pushed

// Functions prototyping to be handled on each Encoder Event
void handleRotate(int8_t rotation) {
  if (rotation > 0)
    rotate_event = 2;  // CW
  else
    rotate_event = 1;  // CCW
}

void handlePressRelease() {
  press_event = 1;
}

void handleLongPressRelease() {
  long_press_event = 1;
}

void setup(void) {

  // Load to the encoder all nedded handle functions here (up to 9 functions)
  versatile_encoder.setHandleRotate(handleRotate);
  versatile_encoder.setHandlePressRelease(handlePressRelease);
  versatile_encoder.setHandleLongPressRelease(handleLongPressRelease);

  u8g2.begin();
  u8g2.enableUTF8Print();
  mui.begin(u8g2, fds_data, muif_list, sizeof(muif_list) / sizeof(muif_t));
  mui.gotoForm(/* form_id= */ 1, /* initial_cursor_position= */ 0);

  Serial.begin(9600);               // initialize serial
  while (!Serial);

  Serial.println("LoRa Dump Registers");
}

void handle_events(void) {
  // 0 = not pushed, 1 = pushed  
  if ( press_event == 1 ) {
    mui.sendSelect();
    is_redraw = 1;
    press_event = 0;
  }

  // 0 = not pushed, 1 = pushed  
  if ( long_press_event == 1 ) {
    mui.sendSelectWithExecuteOnSelectFieldSearch();
    is_redraw = 1;
    long_press_event = 0;
  }
  
  // 0 = not turning, 1 = CW, 2 = CCW
  if ( rotate_event == 1 ) {
    mui.nextField();
    is_redraw = 1;
    rotate_event = 0;
  }
  
  if ( rotate_event == 2 ) {
    mui.prevField();
    is_redraw = 1;
    rotate_event = 0;
  }    
}


void loop(void) {

  /* check whether the menu is active */
  if (mui.isFormActive()) {

    /* update the display content, if the redraw flag is set */
    if (is_redraw) {
      u8g2.firstPage();
      do {
        versatile_encoder.ReadEncoder(); // Do the encoder reading and processing
        mui.draw();
        versatile_encoder.ReadEncoder(); // Do the encoder reading and processing
      } while (u8g2.nextPage());
      is_redraw = 0; /* clear the redraw flag */
    }

    versatile_encoder.ReadEncoder();  // Do the encoder reading and processing
    handle_events();

    

  } else {
    /* the menu should never become inactive, but if so, then restart the menu system */
    mui.gotoForm(/* form_id= */ 1, /* initial_cursor_position= */ 0);

    //Serial.println("loop else");
  }

  /*if (millis() - LastTime >= 1000) {
    //Serial.println(String(millis() - LastTime) + '|' + String(millis()) + '|'+  String(LastTime));
    Serial.println(String(selection2) + '|' + String(LastTime()));
    LastTime = millis();
  }*/
  
}
```

**Reply by ViacheslavK9** [reply]
cleaned my code above little bit, sorry

**Reply by olikraus** [reply]
Side remark: There is a memory issue: `MUIF_U8G2_U16_LIST("ID", &selection, NULL, menu_get_str, menu_get_cnt, mui_u8g2_u16_list_goto_w1_pi)`: The cnt function returns 7 while the get_str function only has 6 menu items.

Moreover, I think you want to do this:
```
  static const char *menu[] = { 
    MUI_21 "Test",
    MUI_22 "Battery",
    MUI_23 "Timer",
    MUI_24 "Activation",
    MUI_1 "Back..",
  };
```

One solution to your problem could be to change `#define MUI_MENU_CACHE_CNT 2` in mui.h to `#define MUI_MENU_CACHE_CNT 1`

**Reply by olikraus** [reply, marked as answer]
Another idea is to reset the position to zero:
``` C++
void clearFormPositionCache( mui_t *mui, uint8_t form_id )
{
  uint8_t i;
  for( i = 0; i < MUI_MENU_CACHE_CNT; i++ )
    if ( mui->menu_form_id[i] == form_id )
      mui->menu_form_cursor_focus_position[i] = 0;
}
```

Then execute this before "sendSelect()"

``` C++
  if ( press_event == 1 ) {
    clearFormPositionCache( ui.getMUI(), 11);   // always reset the position for "form" 11
    mui.sendSelect();
    is_redraw = 1;
    press_event = 0;
  }
```

**Reply by ViacheslavK9** [reply]
one more question, Oliver, now i am using two ways of going back to the previous form:

MUIF_EXECUTE_ON_SELECT_BUTTON("ST", mui_u8g2_btn_exit_wm_fi) + sendSelectWithExecuteOnSelectFieldSearch (look for the MUI_FORM(25) in the example above);
MUIF_BUTTON("G0", mui_u8g2_btn_goto_wm_fi) + MUI_XYAT("G0", 100, 56, 11, " Exit").
Second way "remembers" come back to the cached position, first - does not :(

Second i am using with the forms with limited buttons on the form thats why is ok.
But i need first one for the list forms with dynamic scrollable lists.
I tried to upgrade 1 method with this:

``` C++
//forced exit
uint8_t mui_go_main_menu(mui_t *ui, uint8_t msg) {
  if (msg == MUIF_MSG_FORM_START) {

    clearFormPositionCache(mui.getMUI(), 11);   // always reset the position for "form" 11

    mui.gotoForm(1, selection);

  }
  return 0;
}
```
but it does not work, position one is the result. Inspired by your previous suggest i try another way for method 1:

``` C++
//forced exit
uint8_t mui_go_main_menu(mui_t *ui, uint8_t msg) {
  if (msg == MUIF_MSG_FORM_START) {

    clearFormPositionCache(mui.getMUI(), 11);   // always reset the position for "form" 11

    mui.gotoForm(1, GetFormPositionCache(mui.getMUI(), 1));
  }
  return 0;
}

uint8_t GetFormPositionCache( mui_t *mui, uint8_t form_id )
{
  uint8_t i;
  for( i = 0; i < MUI_MENU_CACHE_CNT; i++ )
  {
    Serial.println(String(mui->menu_form_id[i]) + "|mui_go_main_menu " + mui->menu_form_cursor_focus_position[i]);

    if ( mui->menu_form_id[i] == form_id )
      return(mui->menu_form_cursor_focus_position[i]);
    
  }

}
```
almost done, but wierd! if i came from the 6th line it jump back to the 3rd. If i came from the 2nd it jump back to the first though in Serial i see correct line no 6 (for 6th line - 5 index (because starts from0, right?))

maybe you can help here, stuck, tx..

**Reply by ViacheslavK9** [reply]
I can add jump exit button to the list contained from, but its not nice - while selected - element of the list - covers button, please help.

P.S. also you have this in the documentation of scrollable elemets: 
"Note: Do not use any other fields on the same form than the above mention MUIF fields."

**Reply by olikraus** [reply]
Difficult questions, i only have access to my smartphone at the moment. Maybe I can look into your questions in the next days.

**Reply by ViacheslavK9** [reply]
ok, will be gratefull. I really like your wrap, making boring things cosy 👍

**Reply by olikraus** [reply]
> "Note: Do not use any other fields on the same form than the above mention MUIF fields."

hmm i don't remember why I wrote this. Maybe you can provide a more complete example. It actually should work if the button is placed below the list.

**Reply by ViacheslavK9** [reply]
yes, it works, only visual problems with it (elements covering each other)

WIll try to make my question simple.

You have  mui.gotoForm method which i am using.
If i would be able to work with the second parameter - which is initial cursor position - it will solve my main pain for now :)

Do you have examples how to work with it?
I tried to do plain things like mui.gotoForm(1, 3).
But it still opens form on the first line.
I thought maybe you have hidden indexing of the form elements somewhere, thats why its not working or maybe this feature even not implemented, appreciate your answer here (i am almost a newbye to the C environment cant understand it by myself for now), tx

**Comment by ViacheslavK9** [comment]
Only one thing - if i came from the level above - going back - same way - i dont need to reset menu 11 position,
and i came from the level below - new way - i do need reset.

So i inserted call of your clear function to the go_main_menu function:

```
//forced exit
uint8_t mui_go_main_menu(mui_t *ui, uint8_t msg) {
  if (msg == MUIF_MSG_FORM_START) {

    clearFormPositionCache(mui.getMUI(), 11);   
    mui.sendSelectWithExecuteOnSelectFieldSearch();
  }
  return 0;
}
```
and it works like a charm, thanks!

> Another idea is to reset the position to zero:
...
>

**Reply by olikraus** [reply]
> because starts from0, right?)

correct, the first position has the index 0

**Reply by ViacheslavK9** [reply]
How can i, for example - find code for function mui_GetCurrentFormId?
I found only this:
 int getCurrentFormId(void) { return mui_GetCurrentFormId(&mui); }
and this: int mui_GetCurrentFormId(mui_t *ui);    

but where is the code itself?

**Reply by olikraus** [reply]
Here: https://github.com/olikraus/u8g2/blob/c4f9cd9f8717661c46be16bfbcb0017d785db3c1/csrc/mui.c#L851-L856

**Reply by ViacheslavK9** [reply]
Its out of the topic already, but please, tell me how did you get to the mui.c?

I am coming to this line (mui.h) easy:
![image](https://github.com/olikraus/u8g2/assets/5678922/99baa394-ae72-49a2-8fa2-77921584196e)

 what is the bridge from this point to your (above) code?

**Reply by ViacheslavK9** [reply]
Oliver, just to explain, i don't have problems finding that mui.c file, i am just curious if there is a convenient way to jump to that file from my "mui based" code. Like right click - go to the definition. Probably i want too much :)

**Reply by olikraus** [reply]
hmm... isn't this a problem of the Arduino IDE? I mean, how can I help here? I just write the code itself.

**Reply by ViacheslavK9** [reply]
Maybe, yes, Arduino specific.

But if you know which way i can find out next time that GetCurrentGormID is inside mui.c — it will be a very good hint for me 🙂

**Reply by olikraus** [reply]
linux fgrep could be an option ;-)

**Reply by ViacheslavK9** [reply]
🤷‍♂️
